lib/src/codec/udp/udp_message_decoder.dart

```diff
@@ -0,0 +1,225 @@
+/*
+ * Package : Coap
+ * Author : S. Hamblett <steve.hamblett@linux.com>
+ * Date   : 14/05/2018
+ * Copyright :  S.Hamblett
+ */
+
+import 'package:typed_data/typed_data.dart';
+
+import '../../coap_code.dart';
+import '../../coap_constants.dart';
+import '../../coap_empty_message.dart';
+import '../../coap_message.dart';
+import '../../coap_message_type.dart';
+import '../../coap_option.dart';
+import '../../coap_option_type.dart';
+import '../../coap_request.dart';
+import '../../coap_response.dart';
+import 'datagram_reader.dart';
+import 'message_format.dart' as message_format;
+
+/// Provides methods to parse incoming byte arrays to messages.
+class CoapMessageDecoder {
```

Comment:
We might need an interface for the encoder/decoder, if we have multiple encoders/decoders in the future it might make things a bit easier?

Reply:
I experimented a bit more with the encoder/decoder implementations and noticed that you can simplify the implementation quite drastically. I manage to boil down both serialization and deserialization to one class method each

Therefore we could also consider omitting the interfaces and define serialization/deserialization as plain functions. Alternatively, we could also turn them into (static) class methods or factory constructors on the `CoapMessage` class. Not sure what approach would be better here.

Reply:
Tried out the function-based approach in the latest commit and makes things a bit easier I think. I also implemented the extended token length RFC while I was at it.